playground_server_design.md

```diff
@@ -0,0 +1,130 @@
+
+SQLFlow Playground Server expose an REST API service which enable users to share
+the resources in one playground cluster. Users can take advantage of SQLFlow by
+installing a small plugin on her/his own Jupyter Notebook.
+
+## The Architecture
+
+**SQLFlow Playground Server** is a side-by-side service of our playground cluster.
+Now, it is designed as a http server which receives user login, creates DB
+resource and so on. This server use `cubectl` to manipulate the resource in
+the playground(a k8s cluster). It's in someway the gateway of the playground.
+Like described in below diagram, the interaction of the three subjects could
+be: Clients ask the playground server for some resource. The server authorize
+the client and create the resource on the playground. The client connect to
+the playground and do train/predict tasks using the created resource.
+
+```
+   ----------------run task--------------->
+   |                                      |
+Clients <--> Playground Server <--> Playground
+```
+
+## Supported API
+
+Request url path is composed by the prefix `/api/` and the api name, like:
+
+```url
+    https://playground.sqlflow.tech/api/heart_beat
+```
+This service always uses `https` and only accepts authorized clients
+by checking their certification file.
+
+Currently supported API are:
+| name | method | params | description |
+| - | - | - | - |
+| create_db | POST | {"user_id": "id"} | create a DB for given user, json param |
+| heart_beat| GET  | user_id=id | report a heart beat of given client |
+
+
+## How to Use
+
+### For Service Maintainer
+The maintainer should [provide the playground cluster](../dev.md), and
+bootup a `SQLFlow Playground Server`.  The server should have privillege
+to access the `kubectl` command of the cluster.  To install the server,
+maintainer can use below command:
+```bash
+    mkdir $HOME/workspace
+    cd $HOME/workspace
+    pip install sqlflow_playground
+    mkdir key_store
+    gen_cert.sh server
+    sqlflow_playground --port=50052 \
+      --ca_crt=key_store/ca/ca.crt \
+      --server_key=key_store/server/server.key \
+      --server_crt=key_store/server/server.crt
+```
+In above commands, we first installed the sqlflow playground package
+which carries the main cluster operation logic.  Then, we use the key
+tool to generate a server certification file (Of course, it's not necessary
+if you have your own certification files) which enable us to provide
+`https` service.  Finally, we start the `REST API` service at port 50052.
+
+Our playground service use bi-directional validation.  So, maintainer need
+to generate a certification file for trusted user. Use below command and
+send the generated `.crt` and `.key` file together with the `ca.crt` to
+the user.
+
+```bash
+    gen_cert.sh some_client
+```
+
+### For Users
+To use this service, user should get authorized from the playground's maintainer.
+In detail, user should get `ca.crt`, `client.key` and the `client.crt` file from
```

Comment:
If the user side jupyter notebook only talks to the "playground server", why the client needs these files? The API should be a public HTTPs service. If the cert files are used to identify the user, it should act as an authentication method.

Reply:
In DSW case, we can think the `playground server` as the gateway of `SQLFlow playground`. Users should request for resource from the `playground server`, then talk to the palyground. And these file are used for requesting the `playground server`. We use bi-directional verification. Server use `https`(so has a key/crt pair), and client also provide their certification files too.